blocks/search.py

```diff
@@ -0,0 +1,269 @@
+from blocks.graph import ComputationGraph
+from blocks.utils import unpack, dict_union
+
+floatX = config.floatX
+
+
+def construct_dict(func):
+    def wrapper(self, *inputs):
+        names, outputs = func(self, *inputs)
+        return OrderedDict(zip(names, outputs))
+    return wrapper
+
+
+class BeamSearch(object):
+    """Beam search.
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam.
+    sequence_generator : sequence generator
+        Sequence generator brick.
+    attended : theano variable
+        Theano variable for attended in sequence generator.
+    attended_mask : theano variable
+        Theano variable for attended mask in sequence generator.
+    inputs_dict : dict
+        Dictionary of inputs {name: theano variable for input}. The
+        functions will be constructed with these inputs.
+
+    """
+    def __init__(self, beam_size, sequence_generator, inputs_dict, comp_graph):
+        self.beam_size = beam_size
+        self.sequence_generator = sequence_generator
+        self.inputs_dict = inputs_dict
+        self.generate_names = sequence_generator.generate.outputs
+        self.init_computer = None
+        self.next_computer = None
+        self.attended_computer = None
+        self.initial_state_computer = None
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+        self.context_names = sequence_generator.context_names
+        self.comp_graph = comp_graph
+        self.compiled = False
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beam search."""
+        generator = self.sequence_generator
+
+        contexts = OrderedDict()
+        for name in self.context_names:
+            contexts[name] = unpack(
+                VariableFilter(application=generator.generate,
+                               name='^' + name + '$')(self.comp_graph.variables))
+
+        self.attended_computer = function(self.inputs_dict.values(),
+                                          contexts.values(),
+                                          on_unused_input='ignore')
+
+        initial_states = OrderedDict()
+        for name in self.state_names:
+            initial_states[name] = generator.initial_state(
+                name,
+                self.beam_size,
+                **contexts)
+
+        self.initial_state_computer = function(contexts.values(),
+                                               initial_states.values(),
+                                               on_unused_input='ignore')
+
+        # Define inputs for next values computer
+        cur_variables = OrderedDict()
+        for name, value in initial_states.iteritems():
+            cur_value = tensor.zeros_like(value)
+            cur_value.name = name
+            cur_variables[name] = cur_value
+
+        next_generated = generator.generate(iterate=False,
```

Comment:
I am afraid this is wrong. Sorry for not saying it in the beginning, I read it not attentively enough.

When you give an output to `generate`, it is not used to get the next states. It is potentially used to to compute the next output, but the next output is anyway sampled from the distribution defined by the emitter. You can see that in the [generate code](https://github.com/bartvm/blocks/blob/master/blocks/bricks/sequence_generators.py#L208).

I proposed a workaround in one of my recent commits: you should substitute the output of `readout.emit` with your variable.